### PR TITLE
add native long-press message menu on ios

### PR DESCRIPTION
**Pull request by dnbrwstr**
## Summary

Builds on #6227 to add the native long-press message menu on iOS.

## Changes

- add the custom `tlon-message-context-menu` Expo module
- keep the existing JavaScript action sheet on Android, web, and when a screen reader is active
- drive both surfaces from one shared action model and visibility policy
- keep menu actions and reactions synchronized with the latest message state and permissions

## Review notes

Swift review wanted. The press indication, menu timing, and scale curves are tuned deliberately; motion changes should be treated as design changes.

## How did I test?

- app TypeScript
- message action tests: 25 passed
- SwiftFormat lint
- iOS 26 simulator: exact-head Debug build and install; native menu reactions, actions, drag selection, and surrounding conversation
- Android emulator: existing action-sheet fallback and background conversation visibility

### CODE REVIEW

**Comment by dnbrwstr** [comment]
@codex review

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+                   let presentedAction,
+                   let currentAction,
+                   Self.actionsMatch(presentedAction, currentAction)
+                {
+                    onAction(["id": id])
+                }
+            case let .reaction(value):
+                if selectedReaction == presentedSelectedReaction,
+                   reactions.contains(value)
+                {
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Fixed in 90ad8fb821: the JS host now passes post.id, the native view captures it when presenting, dismisses on rebinding, and requires the same ID before every completion branch.

File: packages/app/ui/hooks/useIsScreenReaderEnabled.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+function subscribe(listener: () => void) {
+  listeners.add(listener);
+
+  if (listeners.size === 1) {
+    void AccessibilityInfo.isScreenReaderEnabled?.()
+      .then(setEnabled)
+      .catch(() => {
+        // Not all platforms implement the query; assume no screen reader.
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/hooks/useIsScreenReaderEnabled.ts:63]: Fixed in 90ad8fb821: the shared store subscribes before querying and versions both events and subscription lifetimes, so stale query results cannot overwrite a newer accessibility event.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+
+    func updateGesture(at windowPoint: CGPoint, isSelecting: Bool) {
+        guard isSelecting else {
+            actionList.updateHighlight(at: nil)
+            reactionBar?.updateHighlight(at: nil)
+            return
+        }
+
+        let localPoint = convert(windowPoint, from: window)
+        actionList.updateHighlight(at: actionList.convert(localPoint, from: self))
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift]: Addressed: drag selection now converts through presentation layers and is disabled while the entrance animation is active.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        return recognizer
+    }()
+
+    private lazy var longPressGestureRecognizer: UILongPressGestureRecognizer = {
+        let recognizer = UILongPressGestureRecognizer(
+            target: self,
+            action: #selector(handleLongPress(_:))
+        )
+        recognizer.minimumPressDuration = Animation.menuDelay
+        recognizer.cancelsTouchesInView = true
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift:58]: Addressed: the native recognizers reject touches inside views whose accessibility identifier starts with ReactionDisplay, preserving the reaction pill long press.

File: packages/app/ui/components/ChatMessage/ChatMessage.tsx [hunk 5]
```diff
@@ -105,62 +108,84 @@ const ChatMessage = ({
...
-        testID="Post"
-        borderRadius={'$m'}
-        overflow="hidden"
-        backgroundColor={
-          isWeb && isHovered ? '$secondaryBackground' : 'transparent'
-        }
+      <MessageContextMenu
+        enabled={Boolean(onLongPress) && usesNativeContextMenu}
+        post={post}
+        postActionIds={postActionIds}
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessage.tsx:121]: Addressed: ChatMessage enables the native host only when the post is not a notice.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        recognizer.delegate = self
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift:59]: Addressed: both long-press recognizers now use the same 40-point allowable movement as the drag-selection dead zone.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+            self.transform = baseTransform
+            self.alpha = baseAlpha
+        }
+    }
+
+    func gestureRecognizer(
+        _: UIGestureRecognizer,
+        shouldRecognizeSimultaneouslyWith _: UIGestureRecognizer
+    ) -> Bool {
+        true
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: simultaneous recognition is limited to the paired long presses, with only the early visual indication allowed to coexist with a scroll-view pan.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+    func dismiss() {
+        dismiss(with: nil)
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        dimView.frame = bounds
+
+        let safeInsets = window?.safeAreaInsets ?? safeAreaInsets
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift:337]: Addressed: presenting the window overlay dismisses the keyboard before menu layout.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+    deinit {
+        if let appDidEnterBackgroundObserver {
+            NotificationCenter.default.removeObserver(appDidEnterBackgroundObserver)
+        }
+    }
+
+    func present(in window: UIWindow) {
+        frame = window.bounds
+        autoresizingMask = [.flexibleWidth, .flexibleHeight]
+        sourceFrame = sourceView?.convert(sourceView?.bounds ?? .zero, to: window) ?? .zero
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift:125]: Addressed: rotation or split-view size changes dismiss the presentation immediately rather than laying out or dismissing toward stale geometry.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+    }
+
+    func present(in window: UIWindow) {
+        // The window-level overlay cannot track React Native's keyboard inset.
+        // Dismiss the keyboard before resolving the menu's available space.
+        window.endEditing(true)
+        frame = window.bounds
+        autoresizingMask = [.flexibleWidth, .flexibleHeight]
+        presentedBoundsSize = window.bounds.size
+        sourceFrame = sourceView?.convert(sourceView?.bounds ?? .zero, to: window) ?? .zero
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift:125]: Addressed: dismissal resolves a fresh frame from the live source view after the keyboard-avoiding layout has moved it.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        let location = recognizer.location(in: window)
+        initialGestureLocation = location
+        gestureExitedDeadZone = false
+
+        let presentation = TlonMessageMenuPresentationView(
+            sourceView: self,
+            restingSourceFrame: indicationRestingFrameInWindow,
+            actions: actions,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift:218]: Addressed: action descriptors participate in presentationKey invalidation, and action dispatch also validates the presented action token against the current model.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    private let showsSeparator: Bool
+
+    init(showsSeparator: Bool) {
+        self.showsSeparator = showsSeparator
+        super.init(frame: .zero)
+        addSubview(button)
+        iconView.contentMode = .scaleAspectFit
+        iconView.isUserInteractionEnabled = false
+        addSubview(iconView)
+        titleLabel.font = .systemFont(ofSize: 17)
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift]: Addressed: action labels now use the preferred body font, scale with content-size category, and rows measure their resulting content.

[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift]: Addressed: action labels now use the preferred body font, scale with content-size category, and rows measure their resulting content.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+            self.reactionBar?.frame = self.targetReactionFrame
+        } completion: { _ in
+            self.isPresenting = false
+            guard !self.isDismissing else {
+                return
+            }
+            self.applyTargetLayout()
+        }
+
+        animateScaleBounce()
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift:182]: Addressed: Reduce Motion uses immediate presentation and dismissal paths without the spring or scale bounce.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        didSet {
+            guard !actions.elementsEqual(oldValue, by: Self.actionsMatch) else {
+                return
+            }
+            presentationView?.dismiss()
+        }
+    }
+
+    var reactions: [String] = []
+    var selectedReaction: String?
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: reaction choices and permission-sensitive tokens participate in presentation invalidation and are revalidated before dispatch.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        _ gestureRecognizer: UIGestureRecognizer,
+        shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer
+    ) -> Bool {
+        // The short press indication and menu long press deliberately overlap.
+        // Every other recognizer, especially the chat list pan, should arbitrate
+        // normally so scrolling cannot continue behind an open menu.
+        (gestureRecognizer === pressIndicationGestureRecognizer
+            && otherGestureRecognizer === longPressGestureRecognizer)
+            || (gestureRecognizer === longPressGestureRecognizer
+                && otherGestureRecognizer === pressIndicationGestureRecognizer)
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: the early press indication may coexist with the list pan so a delayed drag can scroll, while the menu recognizer still competes normally.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 16]
```diff
@@ -0,0 +1,643 @@
...
+
+    private func animatePresentation() {
+        UIView.animate(
+            withDuration: Animation.presentationDuration,
+            delay: 0,
+            usingSpringWithDamping: Animation.presentationDamping,
+            initialSpringVelocity: Animation.presentationVelocity,
+            options: [.beginFromCurrentState, .allowUserInteraction]
+        ) {
+            self.applyPresentedAppearance()
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift:172]: Addressed: action and reaction interaction stays disabled until the presentation animation completes.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+
+            switch selection {
+            case let .action(id):
+                // Props can change while dismissal is finishing. Never dispatch
+                // an action that is no longer present in the latest model.
+                if actions.contains(where: { $0.id == id }) {
+                    onAction(["id": id])
+                }
+            case let .reaction(value):
+                onReaction(["value": value])
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: the emitted reaction token must still match a current reaction descriptor, so permission or selection-state changes during dismissal suppress the stale selection.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+      reactionDetails.self.didReact &&
+      ![...defaultReactions, '🌀'].includes(ownReaction)
+        ? ownReaction
+        : '🌀';
+    return [...defaultReactions, lastReaction];
+  }, [canReact, reactionDetails.self.didReact, reactionDetails.self.value]);
+
+  return (
+    <NativeMessageContextMenu
+      actions={actions}
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx:161]: Addressed: the presentation key now includes the message content key and reaction revision, dismissing a snapshot when its visible post changes.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+            let localPoint = button.convert(point, from: self)
+            button.isHighlighted = button.bounds.contains(localPoint)
+        }
+    }
+
+    func selectedButton(at point: CGPoint) -> (payload: Payload, button: UIButton)? {
+        for (payload, button) in buttons {
+            let localPoint = button.convert(point, from: self)
+            if button.bounds.contains(localPoint) {
+                return (payload, button)
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift:47]: Addressed: manual selection now first requires the point to be inside the visible collection bounds.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+
+            presentationView = nil
+            resetGestureState()
+
+            switch selection {
+            case let .action(id):
+                // Props can change while dismissal is finishing. Never dispatch
+                // an action that is no longer present in the latest model.
+                if contentKey == presentedContentKey,
+                   actions.contains(where: { $0.id == id })
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: action tokens include semantic descriptor and content state, and dispatch requires the current descriptor token to match the presented token.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+      reactions={reactions}
+      selectedReaction={
+        reactionDetails.self.didReact ? reactionDetails.self.value : undefined
+      }
+      contentKey={JSON.stringify([
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx]: Addressed: post reactions now participate in presentationKey, invalidating the open snapshot when visible reaction state changes.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 22]
```diff
@@ -0,0 +1,343 @@
...
+            resetGestureState()
+
+            switch selection {
+            case let .action(id):
+                // Props can change while dismissal is finishing. Require the
+                // selected action to have the same meaning the user saw, not
+                // merely the same stable ID (for example, Hide versus Show).
+                let presentedAction = presentedActions.first { $0.id == id }
+                let currentAction = actions.first { $0.id == id }
+                if contentKey == presentedContentKey,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: action dispatch uses its narrow action token; reaction changes invalidate the visual presentation without becoming the action completion guard.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift [hunk 23]
```diff
@@ -0,0 +1,343 @@
...
+                   let presentedAction,
+                   let currentAction,
+                   Self.actionsMatch(presentedAction, currentAction)
+                {
+                    onAction(["id": id])
+                }
+            case let .reaction(value):
+                if reactions.contains(value) {
+                    onReaction(["value": value])
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageContextMenuView.swift]: Addressed: reaction tokens encode the selected reaction and permission state, and completion compares against the current token before dispatch.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+
+    init(showsSeparator: Bool) {
+        self.showsSeparator = showsSeparator
+        super.init(frame: .zero)
+        addSubview(button)
+        iconView.contentMode = .scaleAspectFit
+        iconView.isUserInteractionEnabled = false
+        addSubview(iconView)
+        titleLabel.font = .preferredFont(forTextStyle: .body)
+        titleLabel.adjustsFontForContentSizeCategory = true
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuAccessoryViews.swift:69]: Addressed: action labels are multiline and word-wrapped, and each row measures the wrapped label to determine its height.

File: packages/app/ui/components/ChatMessage/ChatMessage.tsx [hunk 25]
```diff
@@ -105,62 +108,88 @@ const ChatMessage = ({
...
-        backgroundColor={
-          isWeb && isHovered ? '$secondaryBackground' : 'transparent'
-        }
+      <MessageContextMenu
+        enabled={usesNativeContextMenu}
+        post={post}
+        postActionIds={postActionIds}
+        canReact={canWrite}
+        onReply={handleRepliesPressed}
+        onEdit={handleEditPressed}
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessage.tsx:124]: Addressed: the edit callback accepts the live post supplied by ChatMessage and stores that value in Scroller.

File: apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift [hunk 26]
```diff
@@ -0,0 +1,655 @@
...
+            let point = presentationPoint(windowPoint, in: reactionBar)
+            if let selection = reactionBar.selection(at: point) {
+                dismiss(with: selection)
+                return
+            }
+        }
+
+        let actionPoint = presentationPoint(windowPoint, in: actionList)
+        if let actionId = actionList.selection(at: actionPoint) {
+            dismiss(with: .action(actionId))
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply apps/tlon-mobile/modules/tlon-message-context-menu/ios/TlonMessageMenuPresentationView.swift]: Addressed: both drag update and finish paths reject selection while isPresenting, so unrevealed rows cannot be selected.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+      selectedReaction={
+        reactionDetails.self.didReact ? reactionDetails.self.value : undefined
+      }
+      contentKey={contentKey}
+      reactionKey={JSON.stringify(post.reactions)}
+      alignment={post.authorId === currentUserId ? 'trailing' : 'leading'}
+      previewBackgroundColor={theme.secondaryBackground.val}
+      onAction={(event) => performAction(event.nativeEvent.id)}
+      onReaction={(event) => onEmojiSelect(event.nativeEvent.value)}
+      onMoreReactions={onShowEmojiPicker}
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx]: Addressed: the emoji-picker callback accepts the live post from ChatMessage and uses it as the active message.

File: packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+    descriptor.id === 'viewReactions' ? post.reactions : null,
+  ]);
+}
+
+export function messageContentKey(post: db.Post) {
+  return JSON.stringify([
+    post.content,
+    post.textContent,
+    post.title,
+    post.image,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts:54]: Fixed in 2fc0a256b2. The message presentation snapshot now includes reply count, latest reply time, and reply participant IDs, so an open native preview is invalidated when its visible thread summary changes. Covered by a focused revision-key test.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+
+  const reactionValues = useMemo(() => {
+    if (!canReact) {
+      return [];
+    }
+    const ownReaction = reactionDetails.self.value;
+    const lastReaction =
+      reactionDetails.self.didReact &&
+      ![...defaultReactions, '🌀'].includes(ownReaction)
+        ? ownReaction
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx]: Fixed in 2fc0a256b2. Native reaction slots now normalize legacy shortcodes for display while retaining the stored raw value for selected-state matching and removal. Equivalent default glyphs are deduplicated, with focused tests for legacy and custom reactions.

**Review by latter-bolden — APPROVED** [review]
Tested locally on simulator, this looks great!

It'll be easier to find edge cases once we have it running on actual devices.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+  value: string;
+  token: string;
+};
+
+const NativeMessageContextMenu =
+  requireNativeViewManager<NativeMessageContextMenuProps>(
+    'TlonMessageContextMenu'
+  );
+
+const defaultReactions = ['👍', '❤️', '😂'];
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx]: Fixed in d6989e6276. The native iOS menu now derives its first three reaction slots from emojiUsage through the shared selectFrequentEmojis helper.

File: packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+    post.description,
+    post.cover,
+    post.isDeleted,
+    post.replyCount,
+    post.replyTime,
+    post.replyContactIds,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts:74]: Fixed in d6989e6276. Action tokens now use an action-content revision that excludes reply-summary-only fields, while the presentation key still includes them. Added regression coverage.

File: packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+    post.id,
+    contentKey,
+    descriptor,
+    descriptor.id === 'viewReactions' ? post.reactions : null,
+  ]);
+}
+
+export function messageContentKey(post: db.Post) {
+  return JSON.stringify([
+    post.content,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts:51]: Fixed in d6989e6276. Blob changes now participate in both the visual presentation revision and action-content revision, with regression coverage.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 33]
```diff
@@ -0,0 +1,179 @@
...
+    [reactionSlots]
+  );
+  const moreReactionsToken = canReact
+    ? JSON.stringify([post.id, 'moreReactions'])
+    : undefined;
+  const alignment = post.authorId === currentUserId ? 'trailing' : 'leading';
+  const previewBackgroundColor = theme.secondaryBackground.val;
+  const presentationKey = JSON.stringify([
+    contentKey,
+    actions,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx:150]: Fixed in 67365e5a0e. The native presentation key now includes the message preview presentation state, including author-row visibility, so the menu dismisses instead of retaining a stale snapshot when that layout changes.

File: packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts [hunk 34]
```diff
@@ -0,0 +1,119 @@
...
+    post.isDeleted,
+  ]);
+}
+
+export function messageContentKey(post: db.Post) {
+  return JSON.stringify([
+    messageActionContentKey(post),
+    post.replyCount,
+    post.replyTime,
+    post.replyContactIds,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts:74]: Fixed in 28b20cf5a8. The visual content revision now covers delivery, edit, delete, edited-content, and last-edit state; focused tests verify those status transitions invalidate the native preview.

File: packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts [hunk 35]
```diff
@@ -0,0 +1,126 @@
...
+    post.deliveryStatus,
+    post.editStatus,
+    post.deleteStatus,
+    post.isEdited,
+    post.lastEditContent,
+    post.lastEditTitle,
+    post.lastEditImage,
+    post.replyCount,
+    post.replyTime,
+    post.replyContactIds,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/ChatMessageActions/messageActionModel.ts:74]: Fixed in 5517ef38e3. The native preview revision now includes the rendered thread unread count and notification state, with focused coverage for both changes.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 36]
```diff
@@ -0,0 +1,181 @@
...
+  const moreReactionsToken = canReact
+    ? JSON.stringify([post.id, 'moreReactions'])
+    : undefined;
+  const alignment = post.authorId === currentUserId ? 'trailing' : 'leading';
+  const previewBackgroundColor = theme.secondaryBackground.val;
+  const presentationKey = JSON.stringify([
+    previewKey,
+    contentKey,
+    actions,
+    post.reactions,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx:151]: Fixed in a5d80e69b7. The native preview revision now includes the displayed nickname and avatar for the author and reply participants, so contact synchronization dismisses an open stale snapshot.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 37]
```diff
@@ -0,0 +1,194 @@
...
+    ? JSON.stringify([post.id, 'moreReactions'])
+    : undefined;
+  const alignment = post.authorId === currentUserId ? 'trailing' : 'leading';
+  const previewBackgroundColor = theme.secondaryBackground.val;
+  const contactPresentationKey = useMemo(
+    () =>
+      [post.authorId, ...(post.replyContactIds ?? [])].map((contactId) => {
+        const contact = contactIndex?.[contactId];
+        return [contactId, contact?.nickname, contact?.avatarImage];
+      }),
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx:144]: No change. The native preview is intentionally a point-in-time capture, not a live mirror of every asynchronous descendant. We invalidate for outer-row state that affects action safety or source geometry; wiring every inline mention to dismiss on background contact sync would add broad coupling and unexpectedly close an active menu for a cosmetic update.

File: packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx [hunk 38]
```diff
@@ -0,0 +1,194 @@
...
+    [contactIndex, post.authorId, post.replyContactIds]
+  );
+  const presentationKey = JSON.stringify([
+    previewKey,
+    contentKey,
```
[the remark this answers is not in the document]
Comment (dnbrwstr) [reply packages/app/ui/components/ChatMessage/MessageContextMenu.ios.tsx:149]: No change. Embedded references own independent asynchronous queries, while the native menu preview is intentionally captured at presentation time. Dismissing an active menu whenever any nested reference hydrates or refreshes would require coupling the menu to every embedded renderer and would make normal background loading interrupt interaction. Outer-row/action revisions remain covered by presentationKey.
